atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/BucketAssignerState.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.asts.bucketingthings;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import org.immutables.value.Value;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonSubTypes({
+    @JsonSubTypes.Type(value = ImmutableStart.class, name = BucketAssignerState.Start.TYPE),
+    @JsonSubTypes.Type(value = ImmutableOpen.class, name = BucketAssignerState.Open.TYPE),
+    @JsonSubTypes.Type(value = ImmutableWaitingClose.class, name = BucketAssignerState.WaitingClose.TYPE),
+    @JsonSubTypes.Type(value = ImmutableCloseFromOpen.class, name = BucketAssignerState.CloseFromOpen.TYPE),
+    @JsonSubTypes.Type(value = ImmutableImmediateClose.class, name = BucketAssignerState.ImmediateClose.TYPE),
```

Comment:
For 2, 4 and 5: do these states describe the action that is about to happen, or what has happened? I'd probably use `Opening`, `ClosingFromOpen`, `ImmediateClosing` or similar - if I'm not mistaken these are the stages where we might have partially succeeded the fan-out PUEs